manager/src/Manager.cpp

```diff
@@ -22,47 +18,80 @@ namespace sdk {
     void Manager::start() {
         if (!m_running) {
             m_running = true;
-            xTaskCreate(startRun, "manager", 4096, this, 1, nullptr);
+            auto res = xTaskCreate(Manager::run, "manager", CONFIG_RUN_TASK_STACK_SIZE, nullptr, 1, &m_taskHandle);
+            if (res != pdPASS) {
+                ESP_LOGE(TAG, "Failed to create manager thread, error code: %d", res);
+            }
         }
     }
 
-    void Manager::startRun(void* _this) {
-        Manager* m = static_cast<Manager*>(_this);
-        m->run();
-    }
-
     void Manager::stop() {
         m_running = false;
-    }
+        TaskStatus_t task_status;
 
-    void Manager::run() {
-        ESP_LOGD(TAG, "Starting manager::run()");
+        // Wait for the manager thread to be deleted (deletion happens from inside the tread)
+        do {
+            vTaskDelay(1);
+            vTaskGetInfo(m_taskHandle, &task_status, pdFALSE, eInvalid);
+        } while (task_status.eCurrentState == eDeleted);
 
+        stopAll();
+    }
+
+    void Manager::stopAll() {
         for (auto& entry: m_components) {
-            auto& c   = entry.second.get();
-            auto  res = c.initialize();
-            if (res.has_value()) {
-                ESP_LOGD(TAG, "Initialized component: %s", c.getTag().c_str());
-                entry.first = true;
-            } else {
-                ESP_LOGE(TAG, "Component %s failed to start: %s",
-                         c.getTag().c_str(), res.error().message().c_str());
-                entry.first = false;
+            // entry.first is a bool that describes whether the component is active
+            bool&      componentActive = entry.first;
+            Component& component       = entry.second;
+
+            // If component is not active, don't attempt to stop it
+            if (!componentActive) {
+                break;
             }
+
+            if (component.stop() != Component::Status::STOPPED) {
+                ESP_LOGE(TAG, "Failed to stop component %s on shutdown of manager",
+                         component.getTag().c_str());
+            }
+        }
+    }
+
+    bool Manager::getInitialized() {
```

Comment:
On first glance I looked over the return type and I thought this returned a list of all initialized components. Maybe this could be renamed to Manager::isInitialized() or something.